lib/matplotlib/colorbar.py

```diff
@@ -357,34 +426,119 @@ def draw_all(self):
+            ax.yaxis.set_ticks([], minor=True)
 
         self._set_label()
 
+    def _get_ticker_locator_formatter(self):
+        """
+        This code looks at the norm being used by the colorbar
+        and decides what locator and formatter to use.  If ``locator`` has
+        already been set by hand, it just returns
+        ``self.locator, self.formatter``.
+        """
+        locator = self.locator
+        formatter = self.formatter
+        if locator is None:
+            if self.boundaries is None:
+                if isinstance(self.norm, colors.NoNorm):
+                    nv = len(self._values)
+                    base = 1 + int(nv / 10)
+                    locator = ticker.IndexLocator(base=base, offset=0)
+                elif isinstance(self.norm, colors.BoundaryNorm):
+                    b = self.norm.boundaries
+                    locator = ticker.FixedLocator(b, nbins=10)
+                elif isinstance(self.norm, colors.LogNorm):
+                    locator = ColorbarLogLocator(self)
+                elif isinstance(self.norm, colors.SymLogNorm):
+                    # The subs setting here should be replaced
+                    # by logic in the locator.
+                    locator = ticker.SymmetricalLogLocator(
+                                      subs=np.arange(1, 10),
+                                      linthresh=self.norm.linthresh,
+                                      base=10)
+                else:
+                    if mpl.rcParams['_internal.classic_mode']:
+                        locator = ticker.MaxNLocator()
+                    else:
+                        locator = ColorbarAutoLocator(self)
+            else:
+                b = self._boundaries[self._inside]
+                locator = ticker.FixedLocator(b, nbins=10)
+        _log.debug('locator: %r', locator)
+        return locator, formatter
+
+    def _use_adjustable(self):
+        """
+        Return if we should use an adjustable tick locator or a fixed
+        one.  (check is used twice so factored out here...)
+        """
+        return (self.boundaries is None
+                and self.values is None
+                and ((type(self.norm) == colors.Normalize)
+                    or (type(self.norm) == colors.LogNorm)))
+
     def update_ticks(self):
         """
         Force the update of the ticks and ticklabels. This must be
         called whenever the tick locator and/or tick formatter changes.
         """
         ax = self.ax
-        ticks, ticklabels, offset_string = self._ticker()
-        if self.orientation == 'vertical':
-            ax.yaxis.set_ticks(ticks)
-            ax.set_yticklabels(ticklabels)
-            ax.yaxis.get_major_formatter().set_offset_string(offset_string)
+        # get the locator and formatter.  Defaults to
+        # self.locator if not None..
+        locator, formatter = self._get_ticker_locator_formatter()
+
+        if self._use_adjustable():
```

Comment:
This function would also benefit from the use of `short_axis` and `long_axis` as suggested above.